e2e/security/security.go

```diff
@@ -0,0 +1,292 @@
+func testSecurityUnpriv(t *testing.T) {
+	tests := []struct {
+		name          string
+		image         string
+		action        string
+		expectID      string
+		argv          []string
+		opts          e2e.ExecOpts
+		exit          int
+		expectSuccess bool
+	}{
+		// taget UID/GID
+		{
+			name:          "Set_uid",
+			image:         imagePath,
+			action:        "exec",
+			argv:          []string{"id", "-u"},
+			opts:          e2e.ExecOpts{Security: []string{"uid:99"}},
+			expectID:      "99",
+			exit:          1,
+			expectSuccess: false,
+		},
+		{
+			name:          "Set_gid",
+			image:         imagePath,
+			action:        "exec",
+			argv:          []string{"id", "-g"},
+			opts:          e2e.ExecOpts{Security: []string{"gid:99"}},
+			expectID:      "99",
+			exit:          1,
+			expectSuccess: false,
+		},
+		// seccomp from json file
+		{
+			name:          "SecComp_BlackList",
+			image:         imagePath,
+			action:        "exec",
+			argv:          []string{"mkdir", "/tmp/foo"},
+			opts:          e2e.ExecOpts{Security: []string{"seccomp:./testdata/seccomp-profile.json"}},
+			exit:          1,
+			expectSuccess: false,
+		},
+		{
+			name:          "SecComp_true",
+			image:         imagePath,
+			action:        "exec",
+			argv:          []string{"true"},
+			opts:          e2e.ExecOpts{Security: []string{"seccomp:./testdata/seccomp-profile.json"}},
+			exit:          0,
+			expectSuccess: true,
+		},
+		// capabilities
+		{
+			name:          "capabilities_keep_true",
+			image:         imagePath,
+			action:        "exec",
+			argv:          []string{"ping", "-c", "1", "8.8.8.8"},
+			opts:          e2e.ExecOpts{KeepPrivs: true},
+			exit:          1,
+			expectSuccess: false,
+		},
+		{
+			name:          "capabilities_keep-false",
+			image:         imagePath,
+			action:        "exec",
+			argv:          []string{"ping", "-c", "1", "8.8.8.8"},
+			opts:          e2e.ExecOpts{KeepPrivs: false},
+			exit:          1,
+			expectSuccess: false,
+		},
+		{
+			name:          "capabilities_drop",
+			image:         imagePath,
+			action:        "exec",
+			argv:          []string{"ping", "-c", "1", "8.8.8.8"},
+			opts:          e2e.ExecOpts{DropCaps: "CAP_NET_RAW"},
+			exit:          1,
+			expectSuccess: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run("unpriv "+tt.name, test.WithoutPrivilege(func(t *testing.T) {
+			stdout, stderr, exitCode, err := e2e.ImageExec(t, testenv.CmdPath, tt.action, tt.opts, tt.image, tt.argv)
+
+			lines := bytes.Fields([]byte(stdout))
+
+			if tt.expectSuccess && (exitCode != 0) {
+				t.Log(stderr, err, exitCode)
+				t.Fatalf("unexpected failure running %q: %v", strings.Join(tt.argv, " "), err)
+			} else if !tt.expectSuccess && (exitCode != 1) {
+				t.Log(stderr, err, exitCode)
+				t.Fatalf("unexpected success running %q", strings.Join(tt.argv, " "))
+			}
+
+			//if len(lines) != 1 || string(lines[0]) != "99" {
```

Comment:
note how you are landing here if there was an expectation of failure and the code did in fact fail. In that case I would expect this code to fail. I'm guessing that the check len(lines) == 1 is guarding against two things: that the output does indeed have a line, and that in case of failure we don't try to perform the additional checks. That's two separate bits of logic bundled into a single check.

Maybe:

```
    if tt.expectSuccess && exitCode == 0 {
        if len(lines) != 1 {
            tt.Fatalf("Unexpected output ...")
        }
        if lines[0] != tt.expectID {
            // wrong ID?
            ...
        }
    } else if !tt.expectSuccess && exitCode != 0 {
        // expect failure, command failed
        // clean up, check output, ...
    }
```